### PR TITLE
Add other matrices from Vuduc paper for SpMV

### PR DESCRIPTION
**Pull request by radha-patel**
Added the symmetric matrices from Figure 2 of Performance Bounds and Optimizations for SMVM [here](https://bebop.cs.berkeley.edu/pubs/vuduc2002-smvm-bounds.pdf). 

I included all the matrices that were on SuiteSparse (a few have been removed). Some are not marked as symmetric on SuiteSparse, but I included them anyway - they'll likely pose correctness issues though... let me know if I should just remove them.

Also, is there a better way we want them grouped? I can categorize by structure if that would help.

### CODE REVIEW

**Comment by radha-patel** [comment]
Also - I'm finding that after the loop goes through several matrices, Julia has a segmentation fault. Should we be reallocating /freeing memory somewhere in the test harness or just avoiding running so many matrices consecutively?

**Comment by willow-ahrens** [comment]
Oh no! A few hints for debugging segfaults:
1. Try removing the @inbounds on the Finch kernel and running with --check-bounds=yes to double check that our code is correct.
2. Double check whether we are running out of memory. If we are, let's double check how big our matrices actually are. You can also try using @allocated to track allocations, and double check that each statement allocates at most 10x the size of the matrix (just to double check we aren't densify-ing anywhere).

**Comment by willow-ahrens** [comment]
Also, yes let's only use the ones that suitesparse has declared symmetric. Which ones are not in suitesparse?

**Review by willow-ahrens — CHANGES_REQUESTED** [review]
See comments in the conversation. Thanks Radha for adding these!

**Comment by radha-patel** [comment]
Matrices not in suitesparse: 

- dense1000 - dense matrix 
- lns3937 - fluid flow modeling; found lnsp3937 though, which is also included
- jpwh991 - circuit physics modeling

Also, out of the 37 applicable matrices, only 13 were symmetric on suitesparse

**Comment by radha-patel** [comment]
Also turns out that the code is in fact trying to access an out of bounds index - e.g. `BoundsError: attempt to access 335872-element Vector{Int64} at index [335873]` when running Julia with --check-bounds=yes. I'll investigate this a bit further.

**Comment by radha-patel** [comment]
Lolll the out-of-bounds error was from using the trick to remove the innermost `phase_stop_2 = (min)(phase_stop, A_lvl_2_i)`. As a result of this, when we're in the last column, `A_lvl_2_q += 1` gets called one-too-many times and this is a problem because of the `A_lvl_idx[A_lvl_2_q]` just before the second while block (which deals with the diagonals) if there is a 0 in the bottom right corner of A (i.e. there is no value on the diagonal in that column) as this would be out-of-bounds. 

Old code: 
~~~
function ssymv_finch_kernel_helper(y::Fiber{DenseLevel{Int64, ElementLevel{0.0, Float64, Int64, Vector{Float64}}}}, A::Fiber{DenseLevel{Int64, SparseListLevel{Int64, Vector{Int64}, Vector{Int64}, ElementLevel{0.0, Float64, Int64, Vector{Float64}}}}}, x::Fiber{DenseLevel{Int64, ElementLevel{0.0, Float64, Int64, Vector{Float64}}}}, y_j::Scalar{0.0, Float64})
    @inbounds begin
            y_lvl = y.lvl
            y_lvl_2 = y_lvl.lvl
            y_lvl_val = y_lvl.lvl.val
            A_lvl = A.lvl
            A_lvl_2 = A_lvl.lvl
            A_lvl_ptr = A_lvl_2.ptr
            A_lvl_idx = A_lvl_2.idx
            A_lvl_2_val = A_lvl_2.lvl.val
            x_lvl = x.lvl
            x_lvl_val = x_lvl.lvl.val
            A_lvl.shape == x_lvl.shape || throw(DimensionMismatch("mismatched dimension limits ($(A_lvl.shape) != $(x_lvl.shape))"))
            x_lvl.shape == A_lvl_2.shape || throw(DimensionMismatch("mismatched dimension limits ($(x_lvl.shape) != $(A_lvl_2.shape))"))
            A_lvl.shape == x_lvl.shape || throw(DimensionMismatch("mismatched dimension limits ($(A_lvl.shape) != $(x_lvl.shape))"))
            Finch.resize_if_smaller!(y_lvl_val, A_lvl.shape)
            Finch.fill_range!(y_lvl_val, 0.0, 1, A_lvl.shape)
            for j_6 = 1:A_lvl.shape
                x_lvl_q = (1 - 1) * x_lvl.shape + j_6
                A_lvl_q = (1 - 1) * A_lvl.shape + j_6
                y_lvl_q_2 = (1 - 1) * A_lvl.shape + j_6
                x_lvl_2_val = x_lvl_val[x_lvl_q]
                y_j_val = 0
                A_lvl_2_q = A_lvl_ptr[A_lvl_q]
                A_lvl_2_q_stop = A_lvl_ptr[A_lvl_q + 1]
                if A_lvl_2_q < A_lvl_2_q_stop
                    A_lvl_2_i1 = A_lvl_idx[A_lvl_2_q_stop - 1]
                else
                    A_lvl_2_i1 = 0
                end
                phase_stop = (min)(x_lvl.shape, A_lvl_2_i1, (+)(j_6, -1))
                if phase_stop >= 1
                    i = 1
                    if A_lvl_idx[A_lvl_2_q] < 1
                        A_lvl_2_q = Finch.scansearch(A_lvl_idx, 1, A_lvl_2_q, A_lvl_2_q_stop - 1)
                    end
                    while i <= phase_stop
                        A_lvl_2_i = A_lvl_idx[A_lvl_2_q]
                        phase_stop_2 = A_lvl_2_i
                        #phase_stop_2 = (min)(phase_stop, A_lvl_2_i)
                        #if A_lvl_2_i == phase_stop_2
                            A_lvl_3_val = A_lvl_2_val[A_lvl_2_q]
                            y_lvl_q = (1 - 1) * A_lvl.shape + phase_stop_2
                            x_lvl_q_2 = (1 - 1) * x_lvl.shape + phase_stop_2
                            x_lvl_2_val_2 = x_lvl_val[x_lvl_q_2]
                            y_lvl_val[y_lvl_q] = (+)((*)(A_lvl_3_val, x_lvl_2_val), y_lvl_val[y_lvl_q])
                            y_j_val = (+)((*)(A_lvl_3_val, x_lvl_2_val_2), y_j_val)
                            A_lvl_2_q += 1
                        #end
                        i = phase_stop_2 + 1
                    end
                end
                phase_start_4 = (max)(1, (+)(1, (+)(j_6, -1)))
                phase_stop_4 = (min)(x_lvl.shape, A_lvl_2_i1)
                if phase_stop_4 >= phase_start_4
                    i = phase_start_4
                    if A_lvl_idx[A_lvl_2_q] < phase_start_4
                        A_lvl_2_q = Finch.scansearch(A_lvl_idx, phase_start_4, A_lvl_2_q, A_lvl_2_q_stop - 1)
                    end
                    while i <= phase_stop_4
                        A_lvl_2_i = A_lvl_idx[A_lvl_2_q]
                        phase_stop_5 = (min)(A_lvl_2_i, phase_stop_4)
                        if A_lvl_2_i == phase_stop_5
                            A_lvl_3_val_2 = A_lvl_2_val[A_lvl_2_q]
                            y_lvl_q = (1 - 1) * A_lvl.shape + phase_stop_5
                            y_lvl_val[y_lvl_q] = (+)((*)(A_lvl_3_val_2, x_lvl_2_val), y_lvl_val[y_lvl_q])
                            A_lvl_2_q += 1
                        end
                        i = phase_stop_5 + 1
                    end
                end
                y_lvl_val[y_lvl_q_2] = (+)(y_j_val, y_lvl_val[y_lvl_q_2])
            end
            qos = 1 * A_lvl.shape
            resize!(y_lvl_val, qos)
            (y = Fiber((DenseLevel){Int64}(y_lvl_2, A_lvl.shape)),)
        end
end
~~~

**Comment by radha-patel** [comment]
I tried a ton of different things to try to restrict `A_lvl_2_q` and either prevent it from going out-of-bounds or prevent the `if phase_stop_4 >= phase_start_4` block from executing if it was in fact out-of-bounds, but all of the solutions led to slowdowns.

Ultimately decided on a new approach - strip diagonals from `A` as well to avoid any masks in the SpMV kernel and store them in a separate array `d` and add `d * x` to `y` along with `y_j`:
~~~
@finch_kernel mode=fastfinch function ssymv_finch_kernel_helper(y, A, x, d, y_j)
        y .= 0
        for j = _
            let x_j = x[j]
                y_j .= 0
                for i = _
                    let A_ij = A[i, j]
                        y[i] += x_j * A_ij
                        y_j[] += A_ij * x[i]
                    end
                end
                y[j] += y_j[] + d[j] * x_j
            end
        end
    end)
~~~
Works and is fast (tbh faster than my prior solution!!)

File: spmv/spmv_finch.jl [hunk 1]
```diff
@@ -132,11 +116,13 @@ using Cthulhu
 function spmv_finch(y, A, x) 
     _y = Fiber!(Dense(Element(0.0)), y)
     _A = Fiber!(Dense(SparseList(Element(0.0))))
-    @finch (_A .= 0; for j = _, i = _; if i <= j; _A[i, j] = A[i, j] end end)
+    _d = Fiber!(Dense(Element(0.0)))
+    @finch (_A .= 0; for j = _, i = _; if i < j; _A[i, j] = A[i, j] end end)
+    @finch (_d .= 0; for j = _, i = _; if i == j; _d[i] = A[i, j] end end)
```
Comment (willow-ahrens) [inline spmv/spmv_finch.jl:121]: this can be a single dimensional loop

**Comment by willow-ahrens** [comment]
looks like the missing matrices are named [`HB/lns_3937`](https://sparse.tamu.edu/HB/lns_3937) and [`HB/jpwh_991`](https://sparse.tamu.edu/HB/jpwh_991). I suppose we could also add a dense 1000x1000 matrix.

**Comment by willow-ahrens** [comment]
If you want to use matrixdepot to add a dense matrix, you could use a custom generator, or pick one of the dense ones. You would need to change matrix names to be tuples in the harness though so that you could pass a size argument to the function.

https://matrix-depot.readthedocs.io/en/latest/user.html

File: spmv/run_spmv.jl [hunk 2]
```diff
@@ -32,7 +32,19 @@ parsed_args = parse_args(ARGS, s)
...
+        "Boeing/crystk03",
+        "Nasa/nasasrb",
+        "Rothberg/3dtube",
+        "Simon/raefsky4",
+        "Mulvey/finan512",
+        "Pothen/pwt",
+        "Cote/vibrobox",
+        "HB/saylr4",
+        "Gupta/gupta1"
+    ],
```
Comment (willow-ahrens) [inline spmv/run_spmv.jl:47]: looks like the missing matrices are named [`HB/lns_3937`](https://sparse.tamu.edu/HB/lns_3937) and [`HB/jpwh_991`](https://sparse.tamu.edu/HB/jpwh_991). Let's skip the dense 1000x1000 matrix.

**Review by willow-ahrens — CHANGES_REQUESTED** [review]
Okay, apart from adding those two changes and simplifying the diagonal construction loop, this looks good!

File: spmv/run_spmv.jl (hunk 2 above)
Comment (radha-patel) [reply spmv/run_spmv.jl:49]: Got it - looks like both of these matrices are not symmetric though so I'm going to skip them.

**Pull request merged**